### ISSUE TITLE
Improve customizability of Bop

### ISSUE THREAD

**Issue by lgeiger** [larq/larq#183]
### Feature motivation

Currently when using the Bop optimizer, it will update the weights on all quantized layers:
https://github.com/larq/larq/blob/b32830b4cca4a69ac1daf1d176239971112620e4/larq/optimizers_v2.py#L83-L84

This can be problematic if one wants to use a quantized layer with no `kernel_quantizer` or a quantizer with higher precision.

### Feature description

It would be good to have a more fine grained control over which layers are trained with Bop and which use other precision.

### Feature implementation

One possible implementation would be to add a fake `lq.quantizers.bop` function that doesn't change the forward pass, but marks this kernel so Bop should handle weight updates. This could be achieved by using a specific name scope or adding an attribute that Bop recognises.
An other possibility would be to explicitly pass a list of layers or variables to Bop.
Can you think of a better way to handle this?

**Comment by koenhelwegen** [comment]
I'm not sure if a `quantizer` is the right way to think about this. For me, a `quantizer` implies a map from a real-valued weight to a quantized weight, and so is tightly coupled to the latent-weight approach.

Ideally, the abstraction would make it obvious which approach is used (real-valued weight/latent-weight binary weight/true binary weight), facilitate conversion to an inference graph later on, and be both flexible & transparent in quantization details (bit precision, scaling factors, etc).

One (verbose and probably bad) way to achieve this is adding additional arguments `kernel_quantization_strategy`, `bias_quantization_strategy` etc. to the layers which take values like `None`, `latent`, `binary`; Bop could then target only `true_binary` variables, and we could assert the strategy is `latent` when a `quantizer` is set. We could use scopes to avoid having to set this for each layer individually; however, for models where the higher-precision layers are quantized to 8-bit a mixture of latent weights and true binary weights is actually very common.

**Comment by koenhelwegen** [comment]
Another problem I have with the current API is that Bop has the real-valued optimizer as an attribute. From the users perspective it doesn't make sense to have this hierarchy - Bop & Adam exist next to each other. The current approach also quickly breaks down if you would have more than two optimizers (e.g. Bop for binary, something else for ternary and Adam for latent-weight 8-bit layers).

Not sure if there is a better approach though. Subclassing `keras.model` so that `model.compile` accepts a list of optimizers and determines which variables are targeted by which optimizer would be great but will probably make our life very complicated. A `GroupedOptimizer` class may be an idea if using three different optimizers becomes a real thing, but currently that seems pretty far out.

**Comment by leonoverweel** [comment]
For the optimizer nesting problem, I like your first suggestion of subclassing `keras.model` and creating `larq.model`. From a user's perspective, I think that's the most intuitive API. An added benefit would be that we can move the logic for determining which optimizer (Bop vs. Adam) deals with which weights (binary vs. real) into this `larq.model`, which I think is a more sensible place than having the binary optimizer calling the real optimizer when it thinks a layer is real. What are some of the pitfalls you see that'll make our life complicated @koenhelwegen?

**Comment by koenhelwegen** [comment]
Mainly unintentionally diverging from expected behaviour of the model class elsewhere, e.g. things like callbacks accessing `model.optimizer` like [here](https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/keras/callbacks.py#L1341).

**Comment by leonoverweel** [comment]
@lgeiger and I chatted about this earlier today. Outcome of our discussion:

### Optimizer API

We have the following three choices:

1. Current setup: Bop has an `optimizer` parameter to which we pass a real-valued optimizer. Bop takes care of whether it or the real-valued optimizer trains a given variable.
2. Subclassing `keras.models.Model` so that `model.compile` can take more than one optimizer. Forking the `compile` or `fit` functions would then take care of which optimizer trains which weights.
3. Creating an `OptimizerGroup` with the binary and real optimizers as attributes, which subclasses `tf.keras.optimizers.Optimizer` we passs as optimizer to `model.compile`. This then takes care of which optimizer trains which weights.

We decided to go for 3. for several reasons. One is separation of concerns: the `OptimizerGroup` will be able to take care of the complexities of selecting which optimizer trains which weights, so that each optimizer just has to do "optimizer stuff." It will also be able to take care of things like the following, which currently Bop needs to do:

```python
    def __getattr__(self, name):
        if name == "lr":
            return self.fp_optimizer.lr
        return super().__getattr__(name)
```

Another reason is that, for now, 2. looks like too much work because `model.compile` and `model.fit` do a lot of different things, and updating/maintaining those to work with multiple optimizers would be a lot of work.

###  Binary check

For making `is_binary` more robust, we're going to try setting a `_is_binary` (or similar) attribute on the variables of the layers that need to be optimized by Bop. We'll replace calls to this:

```python
    @staticmethod
    def is_binary(var):
        return "/kernel" in var.name and "quant_" in var.name
```

With just a `hasattr("is_binary")` check. Setting this explicitly should be more robust than checking the generated names of layers.

---

I'll begin with implementing the `OptimizerGroup` and then move to the `is_binary` check.

**Comment by lgeiger** [comment]
Closed via #304

**Comment by leonoverweel** [comment]
Reopening for now because we haven't addressed the `is_binary` part yet.

**Comment by lgeiger** [comment]
Closed via #442
